sql/tests: TestRandomSyntaxFunctions failed

[(sql/tests).TestRandomSyntaxFunctions failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2903615&tab=buildLog) on [release-20.2@0a5debc3d3321d3397796cd9f29be6978ee97bd5](https://github.com/cockroachdb/cockroach/commits/0a5debc3d3321d3397796cd9f29be6978ee97bd5):

```
        
        Stack trace:
    rsg_test.go:728: 1m55s of 5m0s: 558419 executions, 339272 successful
    rsg_test.go:728: 2m0s of 5m0s: 582651 executions, 354029 successful
    rsg_test.go:728: 2m5s of 5m0s: 606859 executions, 368679 successful
    rsg_test.go:728: 2m10s of 5m0s: 630897 executions, 383305 successful
    rsg_test.go:728: 2m15s of 5m0s: 655868 executions, 398413 successful
    rsg_test.go:728: 2m20s of 5m0s: 679462 executions, 412696 successful
    rsg_test.go:728: 2m25s of 5m0s: 703481 executions, 427427 successful
    rsg_test.go:728: 2m30s of 5m0s: 728098 executions, 442300 successful
    rsg_test.go:728: 2m35s of 5m0s: 752272 executions, 457035 successful
    rsg_test.go:728: 2m40s of 5m0s: 776389 executions, 471621 successful
    rsg_test.go:728: 2m45s of 5m0s: 800575 executions, 486365 successful
    rsg_test.go:728: 2m50s of 5m0s: 825944 executions, 501769 successful
    rsg_test.go:728: 2m55s of 5m0s: 849912 executions, 516337 successful
    rsg_test.go:728: 3m0s of 5m0s: 874860 executions, 531484 successful
    rsg_test.go:728: 3m5s of 5m0s: 899286 executions, 546414 successful
    rsg_test.go:728: 3m10s of 5m0s: 923034 executions, 560774 successful
    rsg_test.go:728: 3m15s of 5m0s: 947829 executions, 575957 successful
    rsg_test.go:728: 3m20s of 5m0s: 971497 executions, 590298 successful
    rsg_test.go:728: 3m25s of 5m0s: 996104 executions, 605259 successful
    rsg_test.go:728: 3m30s of 5m0s: 1019895 executions, 619757 successful
    rsg_test.go:728: 3m35s of 5m0s: 1043815 executions, 634186 successful
    rsg_test.go:728: 3m40s of 5m0s: 1067603 executions, 648637 successful
    rsg_test.go:728: 3m45s of 5m0s: 1092331 executions, 663578 successful
    rsg_test.go:728: 3m50s of 5m0s: 1116837 executions, 678538 successful
    rsg_test.go:728: 3m55s of 5m0s: 1141661 executions, 693537 successful
    rsg_test.go:728: 4m0s of 5m0s: 1166291 executions, 708396 successful
    rsg_test.go:752: Crash detected: 
        SELECT st_segmentize('0103000020E6100000010000000D000000FA71E37E7A7D5BC0F01F0529E59C2040C8B64D69A0055EC0A0688BD0204206C0A9F3137BBF0963C0386F66C1A1B140C0243F2D62B50059C0380F188EDB4F30C0ECC6837C0B795CC00EB299010CB850C040CD4D8805C53940DEF896EB818C49C09CC0D63AF72341403CF291EE8EAC35C010A8E39C8C672E4044BCDDABEA4956408CB589963BFE49C0240B07DDF6034C40875846FB121D50C0EAF35670823E5240DE84423A7A9156C01CF58CCECA23514061E9DE42532C62C0D0570CA230FB3740FA71E37E7A7D5BC0F01F0529E59C2040':::GEOGRAPHY, 'NaN':::FLOAT8) 
        
        Stack trace:
    rsg_test.go:728: 4m5s of 5m0s: 1191487 executions, 723770 successful
    rsg_test.go:728: 4m10s of 5m0s: 1215508 executions, 738384 successful
    rsg_test.go:728: 4m15s of 5m0s: 1240078 executions, 753278 successful
    rsg_test.go:728: 4m20s of 5m0s: 1264401 executions, 768093 successful
    rsg_test.go:728: 4m25s of 5m0s: 1288716 executions, 782886 successful
    rsg_test.go:752: Crash detected: 
        SELECT st_segmentize('0102000020E610000008000000FE03D74136AD55C0A04F9C68C2411A4000ACDC41B908D93F527D7F7CCEA452C000635783E28331403B1380A7AC6E50C01C03960522764740802F2C1DA084EF3F6CC0F5F2F7725940EE69E5275FC2554008B8ECDC6F8135C0A097894FD34645400CDBB183FF0546C0A64DDFFE89AF414024A1B603A74164C0C870ED8E8A8D4840':::GEOGRAPHY, 'NaN':::FLOAT8) 
        
        Stack trace:
    rsg_test.go:728: 4m30s of 5m0s: 1313202 executions, 797690 successful
    rsg_test.go:728: 4m35s of 5m0s: 1337630 executions, 812600 successful
    rsg_test.go:728: 4m40s of 5m0s: 1361687 executions, 827157 successful
    rsg_test.go:728: 4m45s of 5m0s: 1387161 executions, 842691 successful
    rsg_test.go:728: 4m50s of 5m0s: 1411808 executions, 857653 successful
    rsg_test.go:728: 4m55s of 5m0s: 1434949 executions, 871636 successful
    rsg_test.go:761: 1457468 executions, 885349 successful
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestRandomSyntaxFunctions003284995
--- FAIL: TestRandomSyntaxFunctions (300.09s)

```

<details><summary>More</summary><p>


```
make stressrace TESTS=TestRandomSyntaxFunctions PKG=./pkg/sql/tests TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestRandomSyntaxFunctions.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/geo/geogfn/segmentize.go
+++ b/pkg/geo/geogfn/segmentize.go
@@ -25,6 +25,9 @@ import (
 // This works by dividing each segment by a power of 2 to find the
 // smallest power less than or equal to the segmentMaxLength.
 func Segmentize(geography geo.Geography, segmentMaxLength float64) (geo.Geography, error) {
+	if math.IsNaN(segmentMaxLength) || math.IsInf(segmentMaxLength, 1 /* sign */) {
+		return geography, nil
+	}
 	geometry, err := geography.AsGeomT()
 	if err != nil {
 		return geo.Geography{}, err

--- a/pkg/geo/geogfn/segmentize_test.go
+++ b/pkg/geo/geogfn/segmentize_test.go
@@ -12,6 +12,7 @@ package geogfn
 
 import (
 	"fmt"
+	"math"
 	"testing"
 
 	"github.com/cockroachdb/cockroach/pkg/geo"
@@ -94,6 +95,16 @@ func TestSegmentize(t *testing.T) {
 			wkt:              "MULTIPOINT((0 0), (1 1))",
 			maxSegmentLength: -1,
 			expectedWKT:      "MULTIPOINT((0 0), (1 1))",
+		},
+		{
+			wkt:              "LINESTRING(0 0, 1 1)",
+			maxSegmentLength: math.NaN(),
+			expectedWKT:      "LINESTRING(0 0, 1 1)",
+		},
+		{
+			wkt:              "LINESTRING(0 0, 1 1)",
+			maxSegmentLength: math.Inf(1),
+			expectedWKT:      "LINESTRING(0 0, 1 1)",
 		},
 	}
 	for _, test := range segmentizeTestCases {

--- a/pkg/geo/geomfn/segmentize.go
+++ b/pkg/geo/geomfn/segmentize.go
@@ -26,6 +26,9 @@ import (
 // between given two-points such that each segment has length less
 // than or equal to given maximum segment length.
 func Segmentize(g geo.Geometry, segmentMaxLength float64) (geo.Geometry, error) {
+	if math.IsNaN(segmentMaxLength) || math.IsInf(segmentMaxLength, 1 /* sign */) {
+		return g, nil
+	}
 	geometry, err := g.AsGeomT()
 	if err != nil {
 		return geo.Geometry{}, err

--- a/pkg/geo/geomfn/segmentize_test.go
+++ b/pkg/geo/geomfn/segmentize_test.go
@@ -12,6 +12,7 @@ package geomfn
 
 import (
 	"fmt"
+	"math"
 	"testing"
 
 	"github.com/cockroachdb/cockroach/pkg/geo"
@@ -104,6 +105,16 @@ func TestSegmentize(t *testing.T) {
 			wkt:              "MULTIPOINT ((0.0 0.0), (1.0 1.0))",
 			maxSegmentLength: -1,
 			expectedWKT:      "MULTIPOINT ((0.0 0.0), (1.0 1.0))",
+		},
+		{
+			wkt:              "LINESTRING(0 0, 1 1)",
+			maxSegmentLength: math.NaN(),
+			expectedWKT:      "LINESTRING(0 0, 1 1)",
+		},
+		{
+			wkt:              "LINESTRING(0 0, 1 1)",
+			maxSegmentLength: math.Inf(1),
+			expectedWKT:      "LINESTRING(0 0, 1 1)",
 		},
 	}
 	for _, test := range segmentizeTestCases {

